orttraining/orttraining/core/graph/pipeline_transformer.cc

```diff
@@ -691,33 +722,362 @@ Status TransformGraphForPipeline(
     ));
   }
 
-  auto fill_node_args = [&](const Graph& graph,
-                            const std::vector<const NodeArg*>& existed_node_args,
-                            std::vector<std::string>& new_node_arg_names,
-                            std::vector<const NodeArg*>& merged_node_args) {
-    merged_node_args.insert(merged_node_args.end(), existed_node_args.begin(), existed_node_args.end());
-    for (auto& name : new_node_arg_names) {
-      merged_node_args.push_back(graph.GetNodeArg(name));
+  ORT_RETURN_IF_ERROR(SetInputsOutputsAndResolve(graph, new_input_names, new_output_names));
+  return Status::OK();
+}
+
+void AddNewNodeArg(Graph& graph,
+                   const std::string& op_name,
+                   onnx::TensorProto_DataType type,
+                   std::vector<NodeArg*>& new_node_args,
+                   std::vector<std::string>& new_names) {
+  auto& event_id = CreateTypedNodeArg(graph, type, op_name);
+  new_names.push_back(event_id.Name());
+  new_node_args.push_back(&event_id);
+}
+
+template <typename T>
+void AddNewNodeArgAndInitializer(Graph& graph,
+                                 const std::string& op_name,
+                                 onnx::TensorProto_DataType type,
+                                 T data,
+                                 std::vector<NodeArg*>& new_node_args,
+                                 std::vector<std::string>& new_names) {
+  AddNewNodeArg(graph, op_name, type, new_node_args, new_names);
+
+  ONNX_NAMESPACE::TensorProto proto_data;
+  proto_data.set_name(new_names.back());
+  proto_data.set_data_type(type);
+
+  switch (type) {
+    case ONNX_NAMESPACE::TensorProto_DataType_BOOL:
```

Comment:
`switch` is not good. Can we deduce its type from `typename T`?

Reply:
the `switch` is inevitable eventhough we deduce the type from `T`. Because we still need to differenciate different calling method (`add_int32_data` vs `add_int64_data`), it will be one more step of deducing and then entering the `switch` statement.

---
In reply to: [429519064](https://github.com/microsoft/onnxruntime/pull/3996#discussion_r429519064) [](ancestors = 429519064)

Reply:
How about this, let's write
```
void set_int32_scalar(int32_t value, ONNX_NAMESPACE::TensorProto& tensor) {
          proto_data.set_data_type(onnx::TensorProto_DataType::INT32);
        proto_data.add_int32_data(value);
}

void set_int64_scalar(int64_t value, ONNX_NAMESPACE::TensorProto& tensor) {
          proto_data.set_data_type(onnx::TensorProto_DataType::INT64);
        proto_data.add_int64_data(value);
}
```
and call them in AddNewScalarNodeArgAndInitializer. Then, `proto_data.set_data_type(type);` and `switch` can be removed.


---
In reply to: [430577195](https://github.com/microsoft/onnxruntime/pull/3996#discussion_r430577195) [](ancestors = 430577195,429519064)

Reply:
Even though we have `set_int32_scalar` and and `set_int64_scalar`, how we going to call them without a type check (i.e. `if` or `switch`)?

We still need to distinguish based on different data types to decide with function to call.

On another note, I am not sure why switch is so bad that we want to avoid it...

---
In reply to: [430690470](https://github.com/microsoft/onnxruntime/pull/3996#discussion_r430690470) [](ancestors = 430690470,430577195,429519064)

Reply:
The type parameter "T" determines which one to call.

---
In reply to: [430724590](https://github.com/microsoft/onnxruntime/pull/3996#discussion_r430724590) [](ancestors = 430724590,430690470,430577195,429519064)

Reply:
That's what I was referring too. Once inside `AddNewNodeArgAndInitializer<T>`, we still need to check T's type to decide which one to call. This check is inevitable.

---
In reply to: [430729278](https://github.com/microsoft/onnxruntime/pull/3996#discussion_r430729278) [](ancestors = 430729278,430724590,430690470,430577195,429519064)